File: Blind 75/03 - Sliding Window/3-longest-substring-without-repeating-characters.py
'''
Given a string s, find the length of the longest substring without repeating characters.
'''
from time import time


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        l = 0
        longest, curr = 0, 0
        for r in range(len(s)):
            while s[r] in seen:
                seen.remove(s[l])
                l += 1
                curr -= 1
            seen.add(s[r])
            curr += 1
            longest = max(longest, curr)

        return longest

    def reference(self, s: str) -> int:
        charSet = set()
        l = 0
        res = 0

        for r in range(len(s)):
            while s[r] in charSet:
                charSet.remove(s[l])
                l += 1
            charSet.add(s[r])
            res = max(res, r - l + 1)
        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.lengthOfLongestSubstring(case))
                else:
                    self.lengthOfLongestSubstring(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = ['abcabcbb', 'bbbbb', 'pwwkew']
    test.quantify(test_cases)

File: Blind 75/03 - Sliding Window/JavaScript/3-longest-substring-without-repeating-characters.js
const { performance } = require('perf_hooks');

class Solution {
  method(s) {
    const seen = new Set();

    let l = 0;
    let res = 0;
    for (let r = 0; r < s.length; r++) {
      // Remove earliest character if seen
      while (seen.has(s[r])) {
        seen.delete(s[l]);
        l += 1;
      }
      seen.add(s[r]);
      res = Math.max(res, seen.size);
    }

    return res;
  }

  reference(s) {
    const set = new Set();
    let l = 0;
    let max = 0;

    for (let r = 0, sl = s.length; r < sl; r++) {
      while (set.has(s[r])) {
        set.delete(s[l]);
        l++;
      }

      set.add(s[r]);
      max = Math.max(max, set.size);
    }
    return max;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = ['abcabcbb', 'bbbbb', 'pwwkew'];
test.quantify(testCases);
